### PR TITLE
Allow key file tags with tuning information like "A#m +50"

### PR DESCRIPTION
**Pull request by daschuer**
This is written by Rapid Evolution

### CODE REVIEW

**Comment by daschuer** [comment]
This PR fixes https://github.com/mixxxdj/mixxx/issues/10991 and https://github.com/mixxxdj/mixxx/issues/10990

**Comment by daschuer** [comment]
This is open since three month without interest to test and review. I will remove it form the 2.3 milestone.

**Comment by ronso0** [comment]
So, this just strips the additional tuning info so Mixxx can recognise and import the base key?

File: src/test/keyutilstest.cpp [hunk 1]
```diff
@@ -142,6 +142,47 @@ TEST_F(KeyUtilsTest, KeyNameNotation) {
...
+            KeyUtils::guessKeyFromText("Ab -50cents"));
+    EXPECT_EQ(mixxx::track::io::key::A_FLAT_MAJOR, // ionian
+            KeyUtils::guessKeyFromText("04B -50cents"));
+    EXPECT_EQ(mixxx::track::io::key::A_FLAT_MAJOR, // ionian
+            KeyUtils::guessKeyFromText("    4b    -50   cents    "));
+    // Mixxx does not allow this but Rapid Evolution
+    // EXPECT_EQ(mixxx::track::io::key::A_FLAT_MAJOR, // ionian
+    //          KeyUtils::guessKeyFromText("    g  #    -    50   cents    "));
+    // EXPECT_EQ(mixxx::track::io::key::A_FLAT_MAJOR, // ionian
+    //          KeyUtils::guessKeyFromText("    g  #    +    50   cents    "));
```
[the remark this answers is not in the document]
Comment (daschuer) [reply src/test/keyutilstest.cpp:179]: Your comment is at the disabled test. Mixxx fails to pars that because of too many blanks. I do not consider the test data as realistic and failing is OK or even desired. 

The tuning is trimmed away, because Mixxx does not make use of it right now. 

The next iteration of making Mixxx tunig aware is to store the tuning information in the database. This is implemented here: 
https://github.com/mixxxdj/mixxx/pull/11096 From this we can introduce the next iteration of making it tuning aware.

[the remark this answers is not in the document]
Comment (Swiftb0y) [reply src/test/keyutilstest.cpp:179]: Yes thats exactly what I mean, if rapid evolution allows / exports data with that many blanks, we should accept it. Whats your reasoning for concluding that its not realistic?

Stripping the actual tuning information is fine for now imo.

[the remark this answers is not in the document]
Comment (daschuer) [reply src/test/keyutilstest.cpp:179]: This is artificial data for a full test coverage of the Rapid Evolution key parser. No one exports the data like that.
Since the Mixxxx parser works differently this test is IMHO pointless for us.

File: src/track/keyutils.cpp [hunk 2]
```diff
@@ -17,7 +17,7 @@ using mixxx::track::io::key::ChromaticKey_IsValid;
 static const QString s_openKeyPattern("^\\s*(1[0-2]|[1-9])([dm])\\s*$");
 
 // Lancelot notation, the numbers 1-12 followed by a (minor) or b (major).
-static const QString s_lancelotKeyPattern("^\\s*(1[0-2]|[1-9])([ab])\\s*$");
+static const QString s_lancelotKeyPattern("^\\s*(1[0-2]|0?[1-9])([ab])\\s*$");
```
Comment (Swiftb0y) [inline src/track/keyutils.cpp:20]: I struggle to understand what that `0?` is doing there? I guess accounting for zero-padded numbers? Wouldn't it make more sense then to be even more promiscous and allow an arbitrary number of preceding 0's instead of just when padded 2 chars?
```suggestion
static const QString s_lancelotKeyPattern("^\\s*0*(1[0-2]|[1-9])([ab])\\s*$");
```

File: src/track/keyutils.cpp [hunk 3]
```diff
@@ -247,8 +247,17 @@ QString KeyUtils::getGlobalKeyText(const Keys& keys, KeyNotation notation) {
...
+ChromaticKey KeyUtils::guessKeyFromText(QString text) {
+    // Remove Shift (Tuning) Information used by Rapid Evolution like: "A#m +50";
+    int shiftStart = text.indexOf('+');
+    if (shiftStart < 0) {
+        shiftStart = text.indexOf('-');
+    }
+    if (shiftStart >= 0) {
+        text = text.left(shiftStart);
+    }
+    const QString trimmed = text.trimmed();
```
Comment (Swiftb0y) [inline src/track/keyutils.cpp:259]: I'd personally favor to change the regex's instead, but this is fine as stopgap solution if you insist.

File: src/track/keyutils.cpp [hunk 4]
```diff
@@ -247,8 +247,17 @@ QString KeyUtils::getGlobalKeyText(const Keys& keys, KeyNotation notation) {
 }
 
 // static
-ChromaticKey KeyUtils::guessKeyFromText(const QString& text) {
-    QString trimmed = text.trimmed();
+ChromaticKey KeyUtils::guessKeyFromText(QString text) {
```
Comment (Swiftb0y) [inline src/track/keyutils.cpp:250]: Why the change from const-ref to by value?

File: src/track/keyutils.cpp (hunk 2 above)
Comment (daschuer) [reply src/track/keyutils.cpp:20]: Yes, this is for the compatibility with the Rapid Evolution "Key Code" notation using always two digits like "01A". This was the easiest way to make it compatible to Mixxx's Lancelot format which does not have the zero padding.

I think I need to add a comment that describes that.  

I do not know any formats with more leading zeros so I do not see a point to allow that. Do you?

File: src/track/keyutils.cpp (hunk 3 above)
Comment (daschuer) [reply src/track/keyutils.cpp:259]: Yes, once we are aware of tuning this needs to be replaced. 

While the tuning issue exists, luckily most tracks are at A = 440 Hz. So this is something that is only reasonable with detector support like in Rapid Evolution.   

A prominent out of tune Track is "Coldplay - The Scientist" which is recorded with an out of tune piano at A = 446 Hz
https://www.reddit.com/r/Coldplay/comments/8tjhwy/nearly_every_song_on_arobth_is_out_of_tune_or_not/

File: src/track/keyutils.cpp (hunk 2 above)
Comment (Swiftb0y) [reply src/track/keyutils.cpp:20]: I don't know any either, still just padding a single 0 instead of arbitrary many seems like a missed opportunity, doesn't it?

File: src/track/keyutils.cpp (hunk 4 above)
Comment (Swiftb0y) [reply src/track/keyutils.cpp:250]: Sure, if you think reassigning in the body is necessary, its no longer const. It could probably made const again though once that code has been replaced.

**Comment by daschuer** [comment]
Done

**Review by Swiftb0y — APPROVED** [review]
Thank you, I guess nobody else is interested in reviewing so I'll go ahead and merge

**Comment by daschuer** [comment]
Thank you.
